plenum/server/consensus/consensus_data_provider.py

```diff
@@ -75,70 +75,72 @@ def stable_checkpoint(self) -> int:
         return self._stable_checkpoint
 
     @property
-    def checkpoints(self) -> List[Checkpoint]:
+    def checkpoints(self) -> SortedListWithKey:
         """
         List of Checkpoints, which has been sent, but not stabilized yet
         """
-        return self._checkpoints.values()
+        return self._checkpoints
 
     def preprepare_batch(self, pp: PrePrepare):
         """
         After pp had validated, it placed into _preprepared list
         """
+        # TODO: move logic checks upper the functions stack
+        if pp in self._preprepared:
```

Comment:
Should we really throw these exceptions there?
On a new view, new primary will re-send (in NEW_VIEW message) PrePrepares that are already ordered, So, Replica Services need to take care not to call this method if they already had a PrePrepare, but call it if it's not yet pre-prepared...

Reply:
I think that we have functionality which flushes this data before NEW_VIEW will be processed